bin.src/convert_gen2_repo_to_gen3.py

```diff
@@ -0,0 +1,212 @@
+                                   CalibKeyHandler)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description=__doc__,
+                                     formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument("instrumentClass", metavar="lsst.obs.CAMERA.INSTRUMENT",
+                        help=("The full import path to the gen3 Instrument class for this camera"
+                              " (e.g. lsst.obs.decam.DarkEnergyCamera)."))
+    parser.add_argument("--gen2root", required=True,
+                        help="Root path of the gen 2 repo to be converted.")
+    parser.add_argument("--gen3root", required=True,
+                        help="Root path of the gen 3 repo to be produced.")
+    parser.add_argument("--skymapName", default=None,
+                        help="Name of the new gen3 skymap (e.g. 'discrete/ci_hsc').")
+    parser.add_argument("--skymapConfig", default=None,
+                        help="Path to skymap config file defining the new gen3 skymap.")
+    parser.add_argument("--calibs", default=None,
+                        help="Path to calibration repo; absolute, or relative to gen2root.")
+    parser.add_argument("-v", "--verbose", action="store_const", dest="logLevel",
+                        default=lsst.log.Log.INFO, const=lsst.log.Log.DEBUG,
+                        help="Set the log level to DEBUG.")
+    parser.add_argument("--calibFilterType", default="physical_filter",
+                        help="physical_filter or abstract_filter as the id in the gen2 calibRegistry.")
+    parser.add_argument("-c", "--config", default=None,
+                        help=("Path to a butler configuration file to be included after "
+                              "the Instrument config overrides are applied."))
+
+    args = parser.parse_args()
+
+    skymapList = [args.skymapName, args.skymapConfig]
+    if not all(x is None for x in skymapList) and not all(x is not None for x in skymapList):
+        parser.error("Must specify both --skymapName and --skymapConfig.")
+
+    return args
+
+
+def configure_translators(instrument, calibFilterType, ccdKey="ccd"):
+    """Configure the gen3 translators so they know the correct instrument name.
+    """
+    # Add instrument to Gen3 data ID if Gen2 contains "visit" or ccdKey.
+    # (Both rules will match, so we'll actually set instrument in the same dict twice).
+    Translator.addRule(ConstantKeyHandler("instrument", instrument.getName()),
+                       instrument=instrument.getName(), gen2keys=("visit",), consume=False)
+    Translator.addRule(ConstantKeyHandler("instrument", instrument.getName()),
+                       instrument=instrument.getName(), gen2keys=(ccdKey,), consume=False)
+
+    # Copy Gen2 'visit' to Gen3 'exposure' for raw only.  Also consume filter,
+    # since that's implied by 'exposure' in Gen3.
+    Translator.addRule(CopyKeyHandler("exposure", "visit"),
```

Comment:
Most of the translator rules will need LSST specializations.  DECam and HSC are easier in a lot of ways because they are the ones that already used something like the dimensions we've standardized on for Gen3.

Reply:
DM-23778